proj-xcode/PowerAuth2/PowerAuthSDK.m

```diff
@@ -1236,46 +1255,130 @@ - (void) unlockBiometryKeysWithContext:(LAContext *)context
+        if (!prompt) {
+            prompt = @"< missing prompt >";
+        }
+    }
+    // Prepare policy based on keychain configuration.
+    LAPolicy policy;
+    if (_keychainConfiguration.biometricItemAccess == PowerAuthKeychainItemAccess_AnyBiometricSetOrDevicePasscode) {
+        // The naming is awkward, but 'LAPolicyDeviceOwnerAuthentication' really means that
+        // we're requesting biometry and the device's passcode
+        policy = LAPolicyDeviceOwnerAuthentication;
+    } else {
+        // In this case, only biometry can be used.
+        policy = LAPolicyDeviceOwnerAuthenticationWithBiometrics;
+    }
+    // Now evaluate the policy
+    [context evaluatePolicy:policy localizedReason:prompt reply:^(BOOL success, NSError * _Nullable error) {
         PowerAuthAuthentication * authentication;
-        NSError * error;
-        // Acquire key to unlock biometric factor
-        OSStatus status;
-        NSData * biometryKey = [self biometryRelatedKeyWithAuthentication:keychainAuthentication status:&status];
-        if (biometryKey) {
-            // The biometry key is available, so create a new PowerAuthAuthentication object preconfigured
-            // with possession+biometry factors.
-            authentication = [PowerAuthAuthentication possessionWithBiometryWithCustomBiometryKey:biometryKey
-                                                                              customPossessionKey:nil];
-            error = nil;
+        if (success) {
+            // The LAContext should be pre-authorized now, so the operation is no longer blocking.
+            // Acquire key to unlock biometric factor
+            NSData * biometryKey = [self biometryRelatedKeyWithAuthentication:keychainAuthentication error:&error];
+            if (biometryKey) {
+                // The biometry key is available, so create a new PowerAuthAuthentication object preconfigured
+                // with possession+biometry factors.
+                authentication = [PowerAuthAuthentication possessionWithBiometryWithCustomBiometryKey:biometryKey
+                                                                                  customPossessionKey:nil];
+                error = nil;
+            } else {
+                // Otherwise report an error depending on whether the operation was canceled by the user.
+                authentication = nil;
+            }
         } else {
-            // Otherwise report an error depending on whether the operation was canceled by the user.
+            // Evaluation failed, we should investigate LAError
             authentication = nil;
-            error = PA2MakeError(status == errSecUserCanceled ? PowerAuthErrorCode_BiometryCancel : PowerAuthErrorCode_BiometryFailed, nil);
+            // Embed an original error
+            NSDictionary * errorInfo = error ? @{ NSUnderlyingErrorKey: error } : nil;
+            if ([error.domain isEqualToString:LAErrorDomain]) {
```

Comment:
On the line above, you're checking if the error is nil but here, you're just accessing it. I'm not sure how objc handles this, but shouldn't the error null check be part of the if?

Reply:
`error.domain` is safe because it leads to `nil`, so `[nil isEqualToString:]` is always NO. On opposite to that, if `error` is not provided (perhaps bug in iOS), then constructing a dictionary with nil value will crash. So that's why there's additional check when creating that `NSDictionary`.

Reply:
In other words, accessing `nil` object is OK, but provide `nil` to API that require some object is not OK.